common/djangoapps/enrollment/views.py

```diff
@@ -301,6 +302,81 @@ def get(self, request, course_id=None):
             )
 
 
+class UnenrollmentView(APIView):
+    """
+        **Use Cases**
+
+            * Unenroll a single user from all courses.
+
+              This command should only be issued from a Jenkins job that retires a user
+              from edX.
+
+        **Example Requests**
+
+            POST /api/enrollment/v1/enrollment {
+                "user": "username12345"
+            }
+
+            **POST Parameters**
+
+              A POST request must include the following parameter.
+
+              * user: The username of the user being unenrolled.
+              This will never match the username from the request, since
+              the request is issued via Jenkins.
+
+        **POST Response Values**
+
+             If the user does not exist, or the user is already unenrolled
+             from all courses, the request returns an HTTP 404 "Does Not Exist"
+             response.
+
+             If an unexpected error occurs, the request returns an HTTP 500 response.
+
+            If the request is successful, an HTTP 200 "OK" response is
+            returned along with a list of all courses from which the user was unenrolled.
+        """
+    authentication_classes = (JwtAuthentication,)
+    permission_classes = (permissions.IsAuthenticated, CanRetireUser)
+
+    def post(self, request):
+        """
+        Unenrolls the specified user from all courses.
+        """
+        # Get the User from the request.
+        username = request.data.get('user', None)
+        if not username:
+            return Response(
+                status=status.HTTP_404_NOT_FOUND,
+                data={
+                    'message': u'The user was not specified.'
+                }
+            )
+        try:
+            # make sure the specified user exists
+            User.objects.get(username=username)
+        except ObjectDoesNotExist:
+            return Response(
+                status=status.HTTP_404_NOT_FOUND,
+                data={
+                    'message': u'The user {} does not exist.'.format(username)
+                }
+            )
+        try:
+            enrollments = api.get_enrollments(username)
+            active_enrollments = [enrollment for enrollment in enrollments if enrollment['is_active']]
+            if len(active_enrollments) < 1:
+                return Response(
+                    status=status.HTTP_404_NOT_FOUND,
```

Comment:
I'm not sure this case should return a 404. It's entirely possible for a learner to be retired that wouldn't have any course enrollments. The POST is saying "Unenroll this learner from all courses." - having no unenrollments to do seems more like a 200 than a 404?

Reply:
That makes total sense to me. I was going off of https://openedx.atlassian.net/browse/EDUCATOR-2603, which states "if unenrolling has succeeded previously, the endpoint will respond the same as if no such user exists."

Reply:
Yeah - *I* actually wrote that text, so I'm willing to admit my mistake! 😈 In the case of this endpoint, it's more general-purpose and doesn't actually change the username/email. Calling the endpoint requests, "Please ensure that this learner gets unenrolled from all their enrolled courses." If a learner has no enrolled courses upon return, whether due to unenrolling or not being enrolled before the call, feels like a success (200) either way. So I think we agree on this point.

Reply:
Cool, I'll update to a 200!